### ISSUE TITLE
xml.parsers.expat.ExpatError: not well-formed on get_KGML

### ISSUE THREAD

**Issue by iquasere** [biopython/biopython#3465]
### Setup

I am reporting a problem with Biopython version, Python version, and operating
system as follows:

```
1.77

3.7.6 (default, Jan  8 2020, 19:59:22)
[GCC 7.3.0]

CPython
Linux-4.15.0-115-generic-x86_64-with-debian-buster-sid
```

### Expected behaviour

Getting the kgml of map ```03013``` of KEGG

### Actual behaviour

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jsequeira/anaconda3/lib/python3.7/site-packages/Bio/KEGG/KGML/KGML_pathway.py", line 89, in get_KGML
    reparsed = minidom.parseString(rough_xml)
  File "/home/jsequeira/anaconda3/lib/python3.7/xml/dom/minidom.py", line 1968, in parseString
    return expatbuilder.parseString(string)
  File "/home/jsequeira/anaconda3/lib/python3.7/xml/dom/expatbuilder.py", line 925, in parseString
    return builder.parseString(string)
  File "/home/jsequeira/anaconda3/lib/python3.7/xml/dom/expatbuilder.py", line 223, in parseString
    parser.Parse(string, True)
xml.parsers.expat.ExpatError: not well-formed (invalid token): line 3, column 30303
```

### Steps to reproduce

```
KGML_parser.read(kegg_get("ko03013", "kgml")).get_KGML()
```

**Comment by peterjc** [comment]
Reproducible here on macOS, using latest Biopython from git:

```python
>>> from Bio.KEGG.REST import kegg_get
>>> from Bio.KEGG.KGML import KGML_parser
>>> KGML_parser.read(kegg_get("ko03013", "kgml")).get_KGML()

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/xxx/miniconda3/lib/python3.7/site-packages/Bio/KEGG/KGML/KGML_pathway.py", line 89, in get_KGML
    reparsed = minidom.parseString(rough_xml)
  File "/Users/xxx/miniconda3/lib/python3.7/xml/dom/minidom.py", line 1968, in parseString
    return expatbuilder.parseString(string)
  File "/Users/xxx/miniconda3/lib/python3.7/xml/dom/expatbuilder.py", line 925, in parseString
    return builder.parseString(string)
  File "/Users/xxx/miniconda3/lib/python3.7/xml/dom/expatbuilder.py", line 223, in parseString
    parser.Parse(string, True)
xml.parsers.expat.ExpatError: not well-formed (invalid token): line 3, column 30301
```

This sounds like invalid XML.

The XML looks OK by eye,

```
>>> print(kegg_get("ko03013", "kgml").read())
<?xml version="1.0"?>
<!DOCTYPE pathway SYSTEM "https://www.kegg.jp/kegg/xml/KGML_v0.7.2_.dtd">
<!-- Creation date: Jul 22, 2020 17:22:10 +0900 (GMT+9) -->
<pathway name="path:ko03013" org="ko" number="03013"
         title="RNA transport"
         image="https://www.kegg.jp/kegg/pathway/ko/ko03013.png"
         link="https://www.kegg.jp/kegg-bin/show_pathway?ko03013">
    <entry id="54" name="path:ko03013" type="map"
        link="https://www.kegg.jp/dbget-bin/www_bget?ko03013">
        <graphics name="TITLE:RNA transport" fgcolor="#000000" bgcolor="#FFFFFF"
             type="roundrectangle" x="108" y="58" width="136" height="25"/>
    </entry>
...
    <relation entry1="383" entry2="292" type="PPrel">
        <subtype name="inhibition" value="--|"/>
        <subtype name="binding/association" value="---"/>
    </relation>
</pathway>

>>> kegg_get("ko03013", "kgml").url
'http://rest.kegg.jp/get/ko03013/kgml'
```

In what might be a separate problem, the XML failed validation at https://validator.w3.org/ - strange it said https://www.kegg.jp/kegg/xml/KGML_v0.7.2_.dtd was unavailable (but it is for me).

**Comment by iquasere** [comment]
Any advice on this? Most KGMLs are unobtainable at this point, was this because of some update on the KEGG API?

**Comment by peterjc** [comment]
Probably a change on the KEGG side, first step is to establish if the XML is actually valid (it should be), which might explain why Expat thinks there is an invalid token.

Do you know much about XML and parsing it?

**Comment by iquasere** [comment]
I know about parsing XMLs, but don't think I am proefficient enough to contribute to BioPython 😅 might give it a try in the next weeks

**Comment by iquasere** [comment]
This is the header of the KGML obtained for ```map00680``` with Biopython
```
<?xml version="1.0" ?>
<!DOCTYPE pathway
  SYSTEM 'http://www.genome.jp/kegg/xml/KGML_v0.7.1_.dtd'>
<!-- Created by KGML_Pathway.py Wed Jan 13 18:37:11 2021 -->
<pathway image="https://www.kegg.jp/kegg/pathway/ko/ko00680.png" link="https://www.kegg.jp/kegg-bin/show_pathway?ko00680" name="path:ko00680" number="680" org="ko" title="Methane metabolism">
```
I get that ```http://www.genome.jp/kegg/xml/KGML_v0.7.1_.dtd```, and for the little I know about XMLs, seems to me this is the guide to the XML schema (for you guys this might be obvious). Using an XML validator online, it says
```
Line 664 Column 29 : Attribute "name" is required and must be specified for element type "subtype".
Line 664 Column 29 : Attribute "value" is required and must be specified for element type "subtype".
```
And many more like this, with all errors pointing to this ```subtype``` tag. Running validation in Python
```
from lxml import etree
parser = etree.XMLParser(dtd_validation=True)
tree = etree.parse("map00680.xml", parser)
```
Gives the exact same problem. Checking the DTD file, it seems ```name``` and ```value``` attributes are required (this is confirmed by the database schema available [here](https://www.genome.jp/kegg/xml/docs/)):
```
<!ELEMENT subtype EMPTY>
<!ATTLIST subtype name      %subtype-name.type;    #REQUIRED>
<!ATTLIST subtype value     %subtype-value.type;   #REQUIRED>
```
 Indeed, the first time ```subtype``` shows up is at line 664:
```
<relation entry1="88" entry2="42" type="ECrel">
    <subtype compound="38"/>
  </relation>
```
where it lacks any ```name``` and ```value``` attributes. However, all of these ocorrences of ```subtype``` belong to another ```subtype```  tag, and in this particular KGML they are always nested inside ```relation``` tags! From the DTD:
```
<!ELEMENT relation (subtype*)>
```
which stands correct! There seems to be some little problem to be solved on the DTD to distinguish between the two types of ```subtype```... BUT! Nothing of this matters, since KEGG released new versions of the maps and the DTD. And in these new maps, the instances of ```subtype``` already come with ```name``` and ```value```. And validating them online gives the good results. So I will just go see how to get these new maps instead, and hopefully implement that in BioPython :D

**Comment by peterjc** [comment]
It sounds like KEGG updated their XML schema, but there was a delay in the computer readable description in the DTD file being updated too.

I think we can close this, and you've learnt a little bit about XML in exploring this as a bonus. Thank you!

**Comment by iquasere** [comment]
I'm sorry, should be reopened, as I see BioPython still downloads the old maps! I'm gonna try contributing to fix that...

**Comment by iquasere** [comment]
I have tracked the error, and it's very weird. The first problem is the update in the DTD version, which is easily solved [this way](https://github.com/iquasere/biopython/commit/f55e0de0a0a1090ad990ee1bf99d1dad6223ec84). But it was still not valid according to the DTD, so I had to do this [brute change](https://github.com/iquasere/biopython/commit/03ef104a7f07ca07275527d9e71a1bda73b89b45). Now, this is bizarre because the KGML that results from ```kegg_get('ko00680', 'kgml').read()``` is perfectly fine, and I cannot, for the life of me, understand where does the ```KGML_parser.read(kegg_get('ko00680', 'kgml')).get_KGML()``` change the XML structure! Still, I'm going to create the pull request, as this "fixes" the problem of ```get_KGML```, at least for this type of attributes.
